### PR TITLE
Add graphastar.Path tests

### PR DESCRIPTION
**Pull request by minkezhang**

### CODE REVIEW

File: lib/hpf/graph.go [hunk 1]
```diff
@@ -27,8 +27,8 @@ var (
 
 // D gets exact cost between two neighboring AbstractNodes.
```
Comment (RyanzpLee) [inline lib/hpf/graph.go:28]: Is cost the amount of time to travel from one node to the next?

File: lib/hpf/graphastar_test.go [hunk 2]
```diff
@@ -23,16 +24,77 @@ var (
...
+		Tiles: []*rtsspb.Tile{
+			{Coordinate: &rtsspb.Coordinate{X: 0, Y: 0}, TerrainType: rtscpb.TerrainType_TERRAIN_TYPE_PLAINS},
+			{Coordinate: &rtsspb.Coordinate{X: 1, Y: 0}, TerrainType: rtscpb.TerrainType_TERRAIN_TYPE_PLAINS},
+			{Coordinate: &rtsspb.Coordinate{X: 2, Y: 0}, TerrainType: rtscpb.TerrainType_TERRAIN_TYPE_PLAINS},
+			{Coordinate: &rtsspb.Coordinate{X: 3, Y: 0}, TerrainType: rtscpb.TerrainType_TERRAIN_TYPE_PLAINS},
+			{Coordinate: &rtsspb.Coordinate{X: 4, Y: 0}, TerrainType: rtscpb.TerrainType_TERRAIN_TYPE_PLAINS},
+			{Coordinate: &rtsspb.Coordinate{X: 5, Y: 0}, TerrainType: rtscpb.TerrainType_TERRAIN_TYPE_PLAINS},
+		},
+		TerrainCosts: []*rtsspb.TerrainCost{
+			{TerrainType: rtscpb.TerrainType_TERRAIN_TYPE_PLAINS, Cost: 1},
```
Comment (RyanzpLee) [inline lib/hpf/graphastar_test.go:96]: Does blocked map mean there are terrain features such as mountains that will block the unit's path and they will have to navigate around? And so I assume simpleLongOpenMap is for a map without obstacles.

**Review by RyanzpLee — APPROVED** [review]
What are the differences between the graph files and the tile files. 

- Are the graph files representing the whole map and every chunk is a node.
- Are the tile files breaking down each map chunk further into multiple tiles?

**Comment by minkezhang** [comment]
> What are the differences between the graph files and the tile files.
> 
> * Are the graph files representing the whole map and every chunk is a node.
> * Are the tile files breaking down each map chunk further into multiple tiles?

I'm using the same terminology as the Botea paper here to reduce confusion --

A Tile object refers to the underlying 1x1 grid in the actual map, e.g. a hill, or an unblocked space.

A cluster is a group of these Tile objects, bounded by a rectangle, e.g. a 3x3 Tile group. The cluster object's main use is to define "transitions" along the cluster borders -- these transitions allow us to build a set of "AbstractNode" objects, which represent gateways into and out of a cluster -- we're essentially saying that all traffic must be routed through these AbstractNodes.

Because the absolute number of gateways in a single cluster is much smaller than the actual number of Tile objects, we have the luxury to pre-compute a totally connected graph between all AbstractNodes in a cluster, for all clusters (in the form of an AbstractEdge). These abstract nodes and edges then make up the abstract Graph object, which is our higher-order "map". We can very quickly find the abstract path in an abstract graph, because all the weights are already cached, and there aren't that many AbstractNode instances to DFS through.

The final part of the puzzle is to compute the path from any point _inside_ a cluster to the cluster gateways -- this is done via A* over the underlying Tile map, but restricted to only being able to consider Tile instances within the same cluster. Because the size of a cluster is much smaller than the total map, this is very fast and may be done on the fly (i.e. when a unit is actually commanded to move).

File: lib/hpf/graph.go (hunk 1 above)
Comment (minkezhang) [reply lib/hpf/graph.go:28]: The AbstractNode cost is the aggregate cost of the path on the underlying TileMap; the cost function is an abstract weight -- it doesn't really have to correspond to any real metrics -- but intuitively, we just want to express that travel from an open space (PLAINS TerrainType) to a blocked space (BLOCKED) should be expensive, whereas travel between two PLAINS Tiles should be cheap.

File: lib/hpf/graphastar_test.go (hunk 2 above)
Comment (minkezhang) [reply lib/hpf/graphastar_test.go:96]: Correct. This is a very simple terrain model, but the idea is there are Tile instance which cannot be occupied (BLOCKED), while all other TerrainTypes may be traversed.

An interesting consequence of this kind of map modeling is that we can't express the concept of "elevation" here -- a 1D map of (PLAINS, BLOCKED, PLAINS) may represent a butte (i.e. only the BLOCKED Tile is elevated) just as well as a cliff (where the BLOCKED Tile and one PLAINS Tile are at the same altitude). If we keep just these two TerrainTypes, then for rendering etc., we will need another map-like data object / field which expresses the actual physical property of the terrain.

File: lib/hpf/graph.go (hunk 1 above)
Comment (RyanzpLee) [reply lib/hpf/graph.go:28]: Ah I see, that makes sense, so we need a way of expressing the cost for movement between different tile types. Thank you

**Pull request merged**